File: solutions/elixir/triangle/2/lib/triangle.ex
defmodule Triangle do
  @doc """
  Return the kind of triangle of a triangle with 'a', 'b' and 'c' as lengths.
  """
  def kind(a, b, c), do: kind(Enum.sort([a, b, c]))

  defp kind([a, _, _]) when a <= 0, do: {:error, "all side lengths must be positive"}
  defp kind([a, b, c]) when a + b <= c, do: {:error, "side lengths violate triangle inequality"}
  defp kind([a, a, a]), do: {:ok, :equilateral}
  defp kind([b, b, _]), do: {:ok, :isosceles}
  defp kind([_, c, c]), do: {:ok, :isosceles}
  defp kind([_, _, _]), do: {:ok, :scalene}
end

File: solutions/python/triangle/2/triangle.py
def is_valid_triangle(f):
    def inner(sides):
        return sum(sides) > 2 * max(sides) and f(len(set(sides)))

    return inner


@is_valid_triangle
def equilateral(number_of_sides):
    return number_of_sides == 1


@is_valid_triangle
def isosceles(number_of_sides):
    return number_of_sides < 3


@is_valid_triangle
def scalene(number_of_sides):
    return number_of_sides == 3
